parsl/app/watcher.py

```diff
@@ -0,0 +1,96 @@
+"""This module provides the bash_watch_app function. This function wraps a bash_app and captures
+any files created by the bash_app. It can be used with or without the file provenance feature.
+
+"""
+from typing import Callable, List, Union
+
+from parsl.app.app import python_app
+from parsl.data_provider.dynamic_files import DynamicFileList
+
+
+@python_app
+def bash_watch(func: Callable,
+               outputs: DynamicFileList,
+               paths: Union[List[str], str] = ".",
+               *args,
+               **kwargs) -> int:
+    """This function wraps a bash_app and captures any files created by the bash_app. This is done
+    by using the watchdog library to watch the specified paths for file creation events. It is most
+    useful when using the file provenance framework.
+
+    Parameters
+    ----------
+    func : callable
+        The bash_app to run.
+    outputs : DynamicFileList
+        The list, partial, or complete, of files to watch for. Can also be empty. When the function
+        completes it will hold the list of files that were created by the bash_app, regardless of
+        what was specified here.
+    paths: string or list
+        The path or paths to watch, recursively, for file creation events. Default is ".".
+    *args : list
+        Arbitrary arguments to pass to the bash_app.
+    **kwargs : dict
+        Arbitrary keyword arguments to pass to the bash_app.
+
+    Returns
+    -------
+    The return code of the wrapped `bash_app`
+    """
+    from watchdog.events import FileSystemEventHandler
```

Comment:
I'm not clear about the semantics here:

this is running concurrently with the bash app, but I think the end effect is that you end up with a list that is the equivalent of the list of files at the end of the bash app minus the list of files already present at the start of the bash app, with any dynamic behaviour during the execution of the app being otherwise irrelevant?

will this give weird results if several watcher-wrapped concurrent apps run watching the same directory?

Reply:
Yes, you end up with the file list you specified. Any dynamic behavior is irrelevant with regard to constructing a list of files created by the app. Yes, it could give "weird"/redundant results if multiple watchers are monitoring the same set of directories. I will add that as a warning/caveat.